extensions/kubernetes/controllerservice/KubernetesControllerService.cpp

```diff
@@ -0,0 +1,205 @@
+    setConfiguration(configuration);
+    initialize();
+}
+
+void KubernetesControllerService::initialize() {
+  if (initialized_) { return; }
+
+  std::lock_guard<std::mutex> lock(initialization_mutex_);
+  ControllerService::initialize();
+  setSupportedProperties({NamespaceFilter, PodNameFilter, ContainerNameFilter});
+  initialized_ = true;
+}
+
+void KubernetesControllerService::onEnable() {
+  std::string namespace_filter;
+  if (getProperty(NamespaceFilter.getName(), namespace_filter) && !namespace_filter.empty()) {
+    namespace_filter_ = std::regex{namespace_filter};
+  }
+
+  std::string pod_name_filter;
+  if (getProperty(PodNameFilter.getName(), pod_name_filter) && !pod_name_filter.empty()) {
+    pod_name_filter_ = std::regex{pod_name_filter};
+  }
+
+  std::string container_name_filter;
+  if (getProperty(ContainerNameFilter.getName(), container_name_filter) && !container_name_filter.empty()) {
+    container_name_filter_ = std::regex{container_name_filter};
+  }
+}
+
+namespace {
+
+struct v1_pod_list_t_deleter {
+  void operator()(v1_pod_list_t* ptr) const noexcept { v1_pod_list_free(ptr); }
+};
+using v1_pod_list_unique_ptr = std::unique_ptr<v1_pod_list_t, v1_pod_list_t_deleter>;
+
+v1_pod_list_unique_ptr getPods(apiClient_t* api_client, core::logging::Logger& logger) {
+  logger.log_info("Calling Kubernetes API listPodForAllNamespaces...");
+  v1_pod_list_unique_ptr pod_list{CoreV1API_listPodForAllNamespaces(api_client,
+                                                                    0,  // allowWatchBookmarks
+                                                                    nullptr,  // continue
+                                                                    nullptr,  // fieldSelector
+                                                                    nullptr,  // labelSelector
+                                                                    0,  // limit
+                                                                    nullptr,  // pretty
+                                                                    nullptr,  // resourceVersion
+                                                                    nullptr,  // resourceVersionMatch
+                                                                    0,  // timeoutSeconds
+                                                                    0)};  // watch
+  logger.log_info("The return code of the Kubernetes API listPodForAllNamespaces call: %ld", api_client->response_code);
+  return pod_list;
+}
+
+}  // namespace
+
+std::vector<KubernetesControllerService::AttributeMap> KubernetesControllerService::getAttributes() {
+  if (!api_client_->getClient()) {
+    logger_->log_warn("The Kubernetes client is not valid, unable to call the Kubernetes API");
+    return {};
+  }
+
+  const auto pod_list = getPods(api_client_->getClient(), *logger_);
+  if (!pod_list) {
+    logger_->log_warn("Could not find any Kubernetes pods");
+    return {};
+  }
```

Comment:
Same here, I think it would be better to return an optional to distinguish the errors from an empty list.

Reply:
done in https://github.com/apache/nifi-minifi-cpp/pull/1257/commits/65411a26509db4cb71466af13284c20231059a57